### ISSUE TITLE
Get rid of the dump/restore temporary directory

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#3755]
When running `rethinkdb dump` and `rethinkdb restore`, users are required to have enough free space in their temporary directory (or whichever directory is passed in `--temp-dir`) for the entire uncompressed database.  This is then tar-gzipped into the final location, which means even more free space is needed (even though we use the `--remove-files` option).

Change `rethinkdb dump` and `rethinkdb restore` to stream the database to/from the destination file directly, probably using the Python `tarfile` module.

Suggested new directory format:
```
export_dir/
  test/
    tbl_a.info
    tbl_a.json.part0
    tbl_a.json.part1
    tbl_b.info
    tbl_b.json
```
* Change `rethinkdb dump` and `rethinkdb restore` to use this directory format.
  * Nice to have: support this directory format in `rethinkdb export` and `rethinkdb import` as well, for consistency.
* When running `rethinkdb dump`, collect chunks separately for each table, then append to the tar file as a new file.
    * This may be done by calling directly into `_export.py` routines and hooking onto the queue of rows read from the database (normally sent to the file writers)
    * Make sure that the file is only renamed to the target file only after the dump is complete
* When running `rethinkdb restore`, stream chunks file-by-file from the source tar file
    * This may be done by calling directly into `_import.py`, and replacing the file reader routines with parsed files from the tarfile.
    * The progress bar would need to be reimplemented in terms of progress through the tar file

**Comment by danielmewes** [comment]
Why do we have to change the directory format?
What's the reason for storing the json file in parts?

Can't we just stream into the tar and stream that through e.g. gz before writing?

**Comment by timmaxw** [comment]
Python's `tarfile` module needs to know the size of a file before it begins writing it to the archive, so we can't stream into the archive. The reason for storing the json files in parts is that we can buffer a part in RAM or `/tmp`, and then add it to the archive. This way we know the file size in advance.

**Comment by danielmewes** [comment]
Ah I see. Seems to be a property of the tar format.

**Comment by larkost** [comment]
I think that this is all possible without too much work. But we need to separate the `dump` and `restore` portions. The latter is a bit easier so I will do that in this message, then a second one for `dump`.

For `restore` this should be really easy to do, since `tarfile` already allows us to access individual files either serially or in any arbitrary way we want. There should be no issues with doing this for multiple processes simultaneously since we are read-only.

The one expiation is if we are streaming through `stdin`, and there we should only be processing it in one thread anyways in most cases, unless we want to make a producer/consumer model to try and keep things moving (and then we are still reading on only one thread).

A side comment here is that we are already creating an extra file (`tar_temp_file_path`) when coming form `sys.stdout`. The `tarfile` module already supports streaming reads with the `r|*`-style formats. I did run into a `UnicodeDecodeError` issue when trying this on Python3.4, so that would need to be worked around (it was in the `zip` portion, so we could probably avoid it).

**Comment by larkost** [comment]
For the `dump` portion we could do three things to drastically reduce/eliminate our on-disk footprint:
1. Rather than writing all of the output files to disk, then packing them into a `tar.gz` file, we could write each one into the archive when it is done, rather than waiting for them all. This would do almost nothing for `dump`s dominated by a single big table, but would improve all other cases.
2. Instead of a `tar.gz` we make the it a `tar` of `gz` files. This means we can stream the `JSON` data straight though the `gzip` module. Then at the end of each file it could be included into the outer `tar` file and the buffer file dropped.
3. If we are willing to only work on a single table/file at a time, it would be easy to implement a `tar` producer that we could stream data into without knowing the size of the file beforehand. We would write out the leading file header for each file and remember the offset for the `size` member in that structure. Then when we finish writing the data for that file `seek` back to that location and write the actual length in before `seek`ing back to the end of the file to continue with the next data file. This would not work as well for streaming to `stdout`, as you would have to buffer each file until it was done, but it would still work.

**Comment by ptgamr** [comment]
I would like to resurrect this issue. Perhaps better for it to be in https://github.com/rethinkdb/rethinkdb-python ?
